node/pkg/telemetry/loki.go

```diff
@@ -142,15 +163,133 @@ func NewLokiCloudLogger(ctx context.Context, logger *zap.Logger, url string, pro
 		lokiLabels[level] = levLabels
 	}
 
+	// Create a buffered channel so the application does not block in the logger.
+	bufferedChan := make(chan api.Entry, bufferSize)
+
+	// Create a local context with a cancel function so we can signal our worker to shutdown when the time comes.
+	workerContext, cancelWorker := context.WithCancel(ctx)
+
+	// Create a channel used by the worker to signal that it has exited.
+	workerExitedC := make(chan struct{}, 1)
+
+	// Kick off the worker to read from the local buffered channel and write to the Loki unbuffered channel.
+	go logWriter(workerContext, localLogger, bufferedChan, workerExitedC, c)
+
 	return &Telemetry{
 		encoder: &guardianTelemetryEncoder{
 			Encoder: zapcore.NewJSONEncoder(zapdriver.NewProductionEncoderConfig()),
 			logger: &ExternalLoggerLoki{
-				c:           c,
-				labels:      lokiLabels,
-				localLogger: localLogger,
+				c:             c,
+				labels:        lokiLabels,
+				localLogger:   localLogger,
+				bufferedChan:  bufferedChan,
+				cancelWorker:  cancelWorker,
+				workerExitedC: workerExitedC,
 			},
 			skipPrivateLogs: skipPrivateLogs,
 		},
 	}, nil
 }
+
+// logWriter is the go routine that takes log messages off the buffered channel and posts them to the Loki client. It can block
+// on the Loki client until our context is canceled, meaning we are shutting down. On shutdown, it tries to flush buffered messages
+// and shutdown the Loki client using a timeout for both actions.
+func logWriter(ctx context.Context, logger *zap.Logger, localC chan api.Entry, workerExitedC chan struct{}, c client.Client) {
+	// pendingEntry is used to save the last log message if the write to Loki is interrupted by the context being canceled. We will attempt to flush it.
+	var pendingEntry *api.Entry
+
+	for {
+		select {
+		case entry, ok := <-localC:
+			if !ok {
+				logger.Error("Loki log writer is exiting because the buffered channel has been closed")
+				return
+			}
+
+			// Write to Loki in a blocking manner unless we are signaled to shutdown.
+			select {
+			case c.Chan() <- entry:
+				pendingEntry = nil
```

Comment:
Should there be a check here that `pendingEntry` is not equal to nil before writing to it?